### DISCUSSION TITLE
WinForms - How do I display DevTools on a form?

### DISCUSSION THREAD

**Discussion by goodwin74** [cefsharp/CefSharp#3317] [Q&A]
How to display DevTools on a form next to the browser. So that DevTools is in the form as a control, and not as a separate window?

**Comment by amaitland** [comment]
There's not enough information to answer this question accurately, please provide more detail. See https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support#when-asking-a-question-please-include-the-following please include basic details whenever you ask a question, version, `WinForms/WPF`, etc.

**Reply by goodwin74** [reply]
I'm sorry. I am using WinForms and cefsharp 86 (Environment: Visual Studio 2019 - Visual C # NET).
I would like to make DevTools part of the form, as implemented in Chrome. So that when you open (DevTools.show ()), not a separate window opens, but appears inside the browser or next to it in the same "Form". How can this be done?

**Reply by amaitland** [reply]
The `WinForms` example in the project source has a demo of hosting `DevTools` in a splitter panel.

The example was updated in https://github.com/cefsharp/CefSharp/pull/2804 you can use this as an example.

A brief sumary.

To host `DevTools` as the child of another control
- https://github.com/cefsharp/CefSharp/blob/e8cf336bf3f2871c400cd01ee090fa42df97adf8/CefSharp.WinForms.Example/BrowserTabUserControl.cs#L518

To manage resizing/close/etc
- https://github.com/cefsharp/CefSharp/blob/e8cf336bf3f2871c400cd01ee090fa42df97adf8/CefSharp.WinForms.Example/Handlers/LifeSpanHandler.cs
- https://github.com/cefsharp/CefSharp/blob/e8cf336bf3f2871c400cd01ee090fa42df97adf8/CefSharp.WinForms.Example/Helper/PopupAsChildHelper.cs

```c#
browser.LifeSpanHandler = new LifeSpanHandler();
```

Quite a bit of this could be extracted out into reusable classes to include in the main project, I just haven't found the time. Contributions welcome.
